PR for #35: implement estimation robust treatment heterogeneity

Constantino-Carreto-Romero opened the pull request:
Closes #35

jorpppp commented on xtevent/_eventols.ado:
```diff
@@ -69,7 +89,17 @@ program define _eventols, rclass
+	* error messages for sun_abraham
+	loc sun_abraham ""
+	if "`cohort'"!="" & "`control_cohort'"!="" {
+		di as text _n "You have specified {bf:cohort} and {bf:control_cohort} options. Event-time coefficients will be estimated with Sun and Abraham's Interaction Weighted Estimator."
```
@Constantino-Carreto-Romero add the year to the reference.

jorpppp commented:
@Constantino-Carreto-Romero In https://github.com/JMSLab/xtevent/commit/6179d2fc456832ab330fbf11ff2319007774f1fa and the log files, it is unclear if the SA and xtevent estimates are converging here, but I think this is because we are comparing estimates that do not control for the omitted variable, and as such are biased. Let's compare estimates from models that control for eta.

jorpppp commented on xtevent/xtevent.sthlp:
```diff
@@ -206,9 +208,17 @@ option requires {help reghdfe} and {help ftools} to be installed. For IV estimat
 unit fixed effects. {opt addabsorb} requires {opt reghdfe}.
 
+{phang}
+{opt cohort(varname)} specifies the variable that identifies the cohort for each unit. {opt cohort} and {opt control_cohort} indicates {cmd:xtevent} to estimate the even-time coefficients with the Interaction Weighted Estimator
```
@Constantino-Carreto-Romero fix typos, emphasize that these options are required for SA.

@jorge-emilio-melendez Can you help to check the writing here?

jorge-emilio-melendez replied:
@jorpppp and @Constantino-Carreto-Romero I found a typo in line 219:

It should be "... to estimate the **event-time** coefficients with the Interaction Weighted Estimator." 
instead of **even-time**

jorpppp replied:
Thanks @jorge-emilio-melendez. @Constantino-Carreto-Romero , can you fix this, please?

jorpppp commented:
@Constantino-Carreto-Romero @jorge-emilio-melendez I left a couple of comments on minor things to check. @jorge-emilio-melendez could you please:

- [ ] Read Sun and Abraham [here](https://www.sciencedirect.com/science/article/pii/S030440762030378X)
- [ ] Then, test the examples that @Constantino-Carreto-Romero  developed to test the SA implementation [here](https://github.com/JMSLab/xtevent/blob/467cca07f153b9b7edde3a157631b099e99350d5/issue35/issue%2035%20examples%20to%20check%20implementation.do)
- [ ] Once you've checked the examples, try to "break" the code, by imputting different options, trying other ways to estimate things, etc... This will help us find bugs.

jorpppp commented:
@Constantino-Carreto-Romero also please keep an eye on the conflicts of this PR with other branches.

Constantino-Carreto-Romero commented:
in

-  https://github.com/JMSLab/xtevent/pull/124/commits/9f3e02707317323740f21b909ad774fe713c4f63 I implemented https://github.com/JMSLab/xtevent/pull/124#discussion_r1024530589
- https://github.com/JMSLab/xtevent/pull/124/commits/8b5202f6120d5e6d90f5ec0316c958571fdb6005 I implemented https://github.com/JMSLab/xtevent/pull/124#issuecomment-1317711362: I changed the examples to check the implementation and they now control for eta. Likewise, at the end of that list of examples, I compared `xtevent` vs `xtevent`'SA, controlling for eta and using different sample sizes, these are the regression tables with [_2000](https://user-images.githubusercontent.com/36095845/202972751-0f37895a-0f74-4ec9-ab34-3e05bf9659b1.png), [_20000](https://user-images.githubusercontent.com/36095845/202972771-294f9a9e-b776-461a-af2b-6c4ca4a56da1.png), and [200000](https://user-images.githubusercontent.com/36095845/202972794-12a96b41-409b-4ce6-8a63-9228856cf5aa.png) observations, respectively.

jorpppp commented:
> @Constantino-Carreto-Romero @jorge-emilio-melendez I left a couple of comments on minor things to check. @jorge-emilio-melendez could you please:
> 
> * [ ]  Read Sun and Abraham [here](https://www.sciencedirect.com/science/article/pii/S030440762030378X)
> * [ ]  Then, test the examples that @Constantino-Carreto-Romero  developed to test the SA implementation [here](https://github.com/JMSLab/xtevent/blob/467cca07f153b9b7edde3a157631b099e99350d5/issue35/issue%2035%20examples%20to%20check%20implementation.do)
> * [ ]  Once you've checked the examples, try to "break" the code, by imputting different options, trying other ways to estimate things, etc... This will help us find bugs.

@jorge-emilio-melendez is still working on this.

jorpppp commented:
> in
> 
> * [9f3e027](https://github.com/JMSLab/xtevent/commit/9f3e02707317323740f21b909ad774fe713c4f63) I implemented [PR for #35: implement estimation robust treatment heterogeneity #124 (comment)](https://github.com/JMSLab/xtevent/pull/124#discussion_r1024530589)
> * [8b5202f](https://github.com/JMSLab/xtevent/commit/8b5202f6120d5e6d90f5ec0316c958571fdb6005) I implemented [PR for #35: implement estimation robust treatment heterogeneity #124 (comment)](https://github.com/JMSLab/xtevent/pull/124#issuecomment-1317711362): I changed the examples to check the implementation and they now control for eta. Likewise, at the end of that list of examples, I compared `xtevent` vs `xtevent`'SA, controlling for eta and using different sample sizes, these are the regression tables with [_2000](https://user-images.githubusercontent.com/36095845/202972751-0f37895a-0f74-4ec9-ab34-3e05bf9659b1.png) (observations per cohort =100), [_20000](https://user-images.githubusercontent.com/36095845/202972771-294f9a9e-b776-461a-af2b-6c4ca4a56da1.png) (observations per cohort =1,000), and [200,000](https://user-images.githubusercontent.com/36095845/202972794-12a96b41-409b-4ce6-8a63-9228856cf5aa.png) (observations per cohort =10,000) observations, respectively.
> * [aae81e2](https://github.com/JMSLab/xtevent/commit/aae81e212575ee7b3e9b96e6102092f40ff9665f), I resolved the conflicts. I ran the test file, again, to check that nothing was affected after resolving the conflicts.
> * [eee749d](https://github.com/JMSLab/xtevent/commit/eee749d61a116d5a7df2bf4bd9ccbbe5fc942954), I added a description for the `saveinteract` suboption (I had missed adding this).

We checked this today. We are happy with the added year reference, and with the comparison between `xtevent` and `xtevent` with the SA estimator.

jorpppp commented:
From @Constantino-Carreto-Romero 's side, this is done. We are missing a couple of checks from @jorge-emilio-melendez and then @jorpppp will take a final look.

jorge-emilio-melendez commented:
> From @Constantino-Carreto-Romero 's side, this is done. We are missing a couple of checks from @jorge-emilio-melendez and then @jorpppp will take a final look.

Hi @jorpppp , I talked to @Constantino-Carreto-Romero and we noticed a package was missing for the SA implementation code to run.

jorpppp commented:
> > From @Constantino-Carreto-Romero 's side, this is done. We are missing a couple of checks from @jorge-emilio-melendez and then @jorpppp will take a final look.
> 
> Hi @jorpppp , I talked to @Constantino-Carreto-Romero and we noticed a package was missing for the SA implementation code to run.

Thanks @jorge-emilio-melendez @Constantino-Carreto-Romero , which package is missing?

Constantino-Carreto-Romero commented:
> Thanks @jorge-emilio-melendez @Constantino-Carreto-Romero , which package is missing?

@jorpppp, I need to add `avar` to the depedency list. I didn't notice this missing before because it was installed in my computer by `EventStudyIntact`, as its dependency.

jorpppp commented:
@Constantino-Carreto-Romero Got it, thanks. I think `avar` is also needed for `reghdfe` anyway, so most people will have it and it does not seem like a burden to add it as a dependency. 
Any other issues that you noticed @jorge-emilio-melendez ?

jorge-emilio-melendez commented:
@jorpppp Everything else seems to run without any issue.

jorpppp commented:
I told @jorge-emilio-melendez about some other ways to try to "break" the code @Constantino-Carreto-Romero. He'll be doing more testing today.

Constantino-Carreto-Romero commented:
In https://github.com/JMSLab/xtevent/pull/124/commits/d3f0881a1069a5077542a35349e7bc18d6b4fe5f, I added `avar` as a dependency.

jorpppp commented:
@Constantino-Carreto-Romero here are some things @jorge-emilio-melendez noticed:

- Without the`cohort` and `control_cohort` options the code estimates the plain event-study without the SA estimator. I think this is OK. @jorge-emilio-melendez notices that this also happens if you leave the options blank. I think this is ok and it is expected behavior, but let me know if you think otherwise.
- Did we incorporate SA estimation in the IV case? Is that pending in another issue?
- @jorge-emilio-melendez did not notice changes in the SA estimates when changing the `impute` option. Is this because our dataset is balanced? Can you test that there are changes when there should be changes with the `impute` option?

jorpppp commented:
@Constantino-Carreto-Romero will create an issue to add SA estimation with IV.

Constantino-Carreto-Romero commented:
@jorpppp @jorge-emilio-melendez 

in https://github.com/JMSLab/xtevent/pull/124/commits/0316070c679e41c024dce1e4e462ad9d492f0423 I added some code snipped to compare the imputation rules. In this case, the example31 dataset has no missing values in the policyvar and it follows staggered adoption, so the different imputation rules have the same effect. [This is the table](https://user-images.githubusercontent.com/36095845/207622214-47993c83-bf78-41c5-9d4b-a079b92dd779.png) to compare them. If I add some inner missing values, I get [this table](https://user-images.githubusercontent.com/36095845/207622430-d7ad4d8f-f60c-492f-93d6-e2b947e01f45.png).

jorpppp commented:
Thanks @Constantino-Carreto-Romero  @jorge-emilio-melendez. Let me now take a last look at this code before we merge the pull.

jorpppp commented on README.md:
```diff
@@ -119,8 +119,21 @@ xtevent ln_w age c.age#c.age ttl_exp c.ttl_exp#c.ttl_exp tenure , ///
+            pol(union) w(3) impute(nuchange) cluster(idcode year) reghdfe ///
+            proxy(wks_work)
+            
+*Interaction Weighted Estimator
```
@Constantino-Carreto-Romero This is a bit unclear, shouldn't we quote this as the Sun and Abraham estimator?

Constantino-Carreto-Romero replied:
I edited it in https://github.com/JMSLab/xtevent/commit/602cf1d6bbdbb4083f9cba70c6b0d47b3f158c51

jorpppp commented on dependency.do:
```diff
@@ -3,3 +3,4 @@ cap ssc install ftools
 cap ssc install ivreghdfe
 cap ssc install ivreg2
 cap ssc install ranktest
+cap ssc install avar
```
@Constantino-Carreto-Romero If I am not mistaken, `ivreg2` uses `avar` for some standard error calculations. If that's the case, then the extra `avar` dependency may be redundant.

On a related note, can we add version numbers for the packages as comments in this file, please?

Constantino-Carreto-Romero replied:
I checked the documentation and ado files and the only `ivreg2`'s dependency is `ranktest`, but `ivreg2` does't install `ranktest` automatically, so it is necessary to keep `ranktest` in the dependency list. On the other hand, `ranktest` does require `avar`, but it doesn't install it automatically, so it is necessary to keep `avar` in the dependecy list. 

In https://github.com/JMSLab/xtevent/commit/c18b7d6db44fc5bc989a941255eca19f739c8d9d, I added the version numbers.

jorpppp commented on issue35/compare sun_abraham vs EventStudyInteract.do:
```diff
@@ -0,0 +1,71 @@
+* compare sun_abraham.ado vs EventStudyInteract
```
@Constantino-Carreto-Romero This file should not be merged to master. In fact, none of the files in the issue folder should be merged to master.

Constantino-Carreto-Romero replied:
In https://github.com/JMSLab/xtevent/commit/bf8eccbe65c5292900df6fe37b18ab0709528096, I deleted the issue folder

jorpppp commented on test/test.do:
```diff
@@ -191,6 +191,17 @@ graph drop g3
 
 graph drop _all
 
+*Interaction Weighted Estimator (Sun and Abraham, 2021) 
```
This is clearer, use this title for readme.md @Constantino-Carreto-Romero

Constantino-Carreto-Romero replied:
@jorpppp, I changed it for "Sun and Abraham Estimator" in https://github.com/JMSLab/xtevent/commit/602cf1d6bbdbb4083f9cba70c6b0d47b3f158c51, but I can still change it for the whole reference "Interaction Weighted Estimator (Sun and Abraham, 2021)".

jorpppp commented on test/test.do:
```diff
@@ -191,6 +191,17 @@ graph drop g3
+gen never_treat=time_of_treat==.
+*Estimate the event-time coefficients with the IW Estimator.
+xtevent y eta , policyvar(z) window(5) vce(cluster i) impute(nuchange) cohort(time_of_treat) control_cohort(never_treat) 
+*Use reghdfe as the underlying estimation command, this produces identical estimations as with EventStudyInteract.
```
@Constantino-Carreto-Romero do not abbreviate as "IW". No need to talk about EventStudyInteract either.

Constantino-Carreto-Romero replied:
I implemented it in https://github.com/JMSLab/xtevent/commit/3f0c7f70a72901a4c458a1e5f43e577ec6eea8b2

jorpppp commented on xtevent/_eventols.ado:
```diff
@@ -159,17 +170,98 @@ program define _eventols, rclass
+		foreach yy of local cohort_list {
+			tempvar cohort_ind resid`yy'
+			qui gen `cohort_ind'  = (`cohort' == `yy') 
+			qui regress `cohort_ind' `nvarlist'  if `touse' & `control_cohort' == 0 [`weight'`exp']  , nocons
```
@Constantino-Carreto-Romero Check if you do this with `_regress`. It's slightly faster.

jorpppp commented on xtevent/_eventols.ado:
```diff
@@ -159,17 +170,98 @@ program define _eventols, rclass
+		qui mat accum `XX' = `nvarlist' if  `touse' & `control_cohort' == 0 [`weight'`exp'], nocons
+		mat `Sxx' = `XX'*1/r(N) 
+		mat `Sxxi' = syminv(`Sxx') 
+		qui avar (`nresidlist') (`nvarlist')  if `touse' & `control_cohort' == 0 [`weight'`exp'], nocons robust
```
@Constantino-Carreto-Romero Do we have a check for whether `avar` is installed earlier in the code, or in `xtevent.ado` ?

Constantino-Carreto-Romero replied:
I implemented it in https://github.com/JMSLab/xtevent/commit/5b93fc240e456eac1a3d781eaeeaee3e8537bbf3. Then, I ran some examples.

jorpppp commented on xtevent/_eventols.ado:
```diff
@@ -159,17 +170,98 @@ program define _eventols, rclass
+		// The scaling factor is 1/N for N the obs of cross-sectional units
+		// But here estimates are on the panel, which is why it is 1/NT instead
+		// Should cancel out for balanced panel, but unbalanced panel is a TODO
+		
```
@Constantino-Carreto-Romero What should be the proper scaling here? Can you please check SA's code again to see if they changed this?

jorpppp replied:
@Constantino-Carreto-Romero thanks. Please add this info about the scaling factor in page 50 and the last code update date as a comment here, to make sure we don't forget when we revisit this.

Constantino-Carreto-Romero replied:
In https://github.com/JMSLab/xtevent/commit/a15a73e40baf57bd02e669253733a4458b9216d5 and https://github.com/JMSLab/xtevent/commit/e28ed455b3ff8f10fdf34466500eb1d7111eff56 I implemented.

jorpppp commented on xtevent/_eventols.ado:
```diff
@@ -159,17 +170,98 @@ program define _eventols, rclass
+				local cohort_rel_varlist "`cohort_rel_varlist' `n`n`l''_`yy''"
+				if "`saveint'"!=""{
+					loc lnumber=substr("`l'",7,2)
+					qui gen _interact_`lnumber'_c`yy' = `n`n`l''_`yy''
```
@Constantino-Carreto-Romero This hardcoded substring length makes me nervous, how are we sure that this will always be the proper length?

(I am aware this is SA's code, but might as well check it)

Constantino-Carreto-Romero replied:
I added that code for the interaction variables to be saved when the suboption `saveinteract` is specified. It substrings on the _k* variables generated in `_eventgenvars` and whose names are stored in the local `included`. The _k* are for instance `_k_eq_m4 _k_eq_m3 _k_eq_m2 _k_eq_p0 _k_eq_p1 _k_eq_p2 _k_eq_p3 _k_eq_p4`, so it will substring `m4 m3 ... p3 p4`. In that code stage, the _k* variables will always have that name structure . It can only be modified later in `_eventols` with an added prefix by the option `savek`.

jorpppp commented on xtevent/_eventols.ado:
```diff
@@ -389,14 +590,17 @@ program define _eventols, rclass
 		cap confirm var __k
 		if !_rc qui drop __k
 		if "`methodt'"=="ols" qui drop _ttrend
+		if "`saveint'"!="" qui drop _interact*
```
@Constantino-Carreto-Romero I remember we have an error message somewhere to reserve the _k variable names for the program in case the user has some variables that start with _k already. We should probably add a similar check for _interact as well.

jorpppp commented on xtevent/xtevent.sthlp:
```diff
@@ -72,7 +74,7 @@ See {help xteventtest} for hypothesis testing after estimation and {help xtevent
 {cmd: xtevent} estimates the effect of a policy variable of interest on a dependent variable using a panel event
 study design. Additional control variables can be included in {it:varlist}. The command allows for estimation when a pre-trend is present using
-the instrumental variables estimator of Freyaldenhoven et al. (2019). {p_end}
+the instrumental variables estimator of Freyaldenhoven et al. (2019). It also allows to use the Interaction Weighted Estimator of Sun and Abraham (2021).{p_end}
```
@Constantino-Carreto-Romero "It also allows estimation in settings with heterogeneous effects by cohort using the Interaction Weighted ..."

Constantino-Carreto-Romero replied:
I implemented it in https://github.com/JMSLab/xtevent/commit/cb067f397837feaeeef0495750f70c3a25bf43a8

jorpppp replied:
@Constantino-Carreto-Romero Please fix the typo.

Constantino-Carreto-Romero replied:
I removed the extra space in https://github.com/JMSLab/xtevent/commit/22b3ea79edf6e8d30fa3c7e854cc00ce4c07da29

jorpppp commented on xtevent/xtevent.sthlp:
```diff
@@ -192,13 +194,17 @@ trend (_ttrend) to the regression. {opt (gmm)} uses the GMM to compute the trend
 
+{phang2}
+{opt saveint:eract} saves interaction variables if {opt cohort} and {opt control_cohort} are specified. {opt noe:stimate} and {opt saveint:eract} cannot be specified simultaneously.
+
```
@Constantino-Carreto-Romero Document that `avar` is needed for the SA options.

Constantino-Carreto-Romero replied:
I implemented it in https://github.com/JMSLab/xtevent/commit/c0e95d7073797f719a264370a5b700095fcb2785

jorpppp replied:
@Constantino-Carreto-Romero Some typos in options vs commands.

Constantino-Carreto-Romero replied:
I changed it in https://github.com/JMSLab/xtevent/commit/c9ae7ac9c4c0ef04ca984ea53c986a8d5cb85d9d

jorpppp commented on xtevent/xtevent.sthlp:
```diff
@@ -209,9 +215,17 @@ option requires {help reghdfe} and {help ftools} to be installed. For IV estimat
+ proposed by Sun and Abraham (2021).
+
+{phang}
+{opt control_cohort(varname)} specifies the binary variable that identifies the control_cohort. {opt cohort} and {opt control_cohort} indicates {cmd:xtevent} to estimate the even-time coefficients with the Interaction Weighted Estimator
```
@Constantino-Carreto-Romero "... the binary variable that identifies the control cohort" (no underscore)

jorpppp reviewed (changes requested):
@Constantino-Carreto-Romero Thanks. I left a few comments in the code, we need a few fixes before merging the PR.

jorpppp commented:
Summary: In this PR we finalized the implementation of Sun and Abraham's estimator.
